RC 82: Manuscript Cases and other changes (#87)
highfidelity/hifi-docs-sphinx: docs/source/create/avatars/custom-animations.rst

Customize Avatar Animations
#####################################

> Your avatar can express custom animations such as dancing, juggling, or waving. You can do this by replacing or overriding the standard animations. All avatars in High Fidelity use the built in standard set of animations by default. These animations are generic and do not change based on your avatar. 

Reviewer: ```suggestion
You can express yourself by overriding High Fidelity's standard set of animations with your own custom animations such as dancing, juggling, or waving. Any custom animations you set up will be independent to each avatar you own and wear.
```

> .. note:: We often update our process to import custom animations for avatars, to make it more user friendly. Keep this in mind before customizing your avatar's animations.

Reviewer: ```suggestion
.. note:: We often update our process for importing custom animations to make it more user friendly. As you develop custom animations, keep in mind that you may need to modify them in the future as our custom animation support continues to improve.
```

.. contents:: On This Page
    :depth: 2

----------------------------------
Prerequisites
----------------------------------

> Consider familiarizing yourself with these terms.

Reviewer: ```suggestion
As we delve deeper into creating custom animations, we may use terminology that you are unfamiliar with. Here are some terms you might come across:
```

+--------------------------+--------------------------------------------------------------------------------------------------+

----------------------------------
Prepare Your Custom Animation
----------------------------------

Before you replace the existing standard animations, you need to prepare your custom animation file. Use our :doc:`Avatar Standards Guide <avatar-standards>` and keep the following guidelines in mind:

- Animations must have standard joint names for High Fidelity.
- Animations must have standard joint orientations (y down the bone).
- Key frames must have key frames for every joint at a uniform interval of 30 frames per second.
- Locomotion animation phase has the left ankle in passing position on the first frame. Try to match this phase if you want your locomotion animation to blend with the default set.

Once you create your animation:

1. Export your animation from the external tool of your choice as an FBX file. 
2. Upload your animation FBX file to a cloud server and copy the URL. 

-----------------------------------
Replace Standard Animations
-----------------------------------

> You can have your avatar express custom animations in different ways:

Reviewer: ```suggestion
You can have your avatar use your custom animations by overriding the default animations. There are two different ways to do this:
```

.. note:: This process to replace an existing animation will take complete control of all avatar joints. Inverse Kinematics of the hands and head of HMD users will be disabled. 

You can also override an existing animation role mapping:

1. Use `MyAvatar.getAnimationRoles <https://apidocs.highfidelity.com/MyAvatar.html#.getAnimationRoles>`_ to view the list of roles for the current avatar. 
2. You can replace the animation for each role with a custom animation (FBX file) using `MyAvatar.overrideRoleAnimation <https://apidocs.highfidelity.com/MyAvatar.html#.overrideRoleAnimation>`_.

> We've listed the animation roles and their description. These are frequently updated. We recommend using ``MyAvatar.getAnimationRoles`` to get the latest animation roles being used. The standard animation FBX files for these roles can be found in the High Fidelity source code repository on `GitHub <https://github.com/highfidelity/hifi/tree/master/interface/resources/avatar/animations>`_.

Reviewer: ```suggestion
We've listed the animation roles and their description. These are frequently updated, so we recommend using ``MyAvatar.getAnimationRoles`` to get the latest animation roles before continuing. The standard animation FBX files for these roles can be found in the High Fidelity source code repository on `GitHub <https://github.com/highfidelity/hifi/tree/master/interface/resources/avatar/animations>`_.
```

| LANDRUN                                   | Running land.                                                        |
+-------------------------------------------+----------------------------------------------------------------------+

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Create a Custom Avatar Animation JSON file
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> You can edit or replace the existing Avatar Animation JSON file to override the standard animations. 

Reviewer: ```suggestion
If you're not comfortable using a script, you can edit or replace the existing Avatar Animation JSON file to override the standard animations. 
```

.. note:: If you create a custom JSON file for your avatar's animations, you will not inherit any updates made to the standard animations' JSON file. You can perform a text merge to the latest version at any time.

The JSON file shows which animation role is mapped to which animation FBX file. You can replace standard animation FBX files with your custom animation's FBX files. Or, you can write a new JSON file with the new mappings for each animation role. 

To replace standard animations:

1. Upload your custom JSON file to a cloud server and copy the URL.
2. In Interface, pull up your HUD or Tablet and go to **Avatar**.
3. Click on the Settings icon on the top-right corner. 

> 4. Under 'Avatar animation JSON', paste the URL for your JSON file. 

Reviewer: ```suggestion
4. Under 'Avatar Animation JSON', paste the URL for your JSON file. 
```

> .. image:: _images/avatar-anim.png

Reviewer: I don't think this image is needed.

Author: Removing it.

> .. image:: _images/avatar-anim.png

Author: ```suggestion

```

^^^^^^^^^^^^^^^^

> 

Reviewer: Add note to the section about changing your FST file (above) that if it is uploaded to our servers/Marketplace, they'll need to update it using the Avatar Packager.

Author: Done.
